docs/internal/GeoffsRecoilAssessment.md

```diff
@@ -0,0 +1,176 @@
+- Only introduce the dependency on recoil where required. This will generally be at higher level container components level.
+- It is perfectly OK to leverage pub/sub event busses where dependency on shared state is awkward across component hierarchy siblings.
+- It is OK to use the Context API to avoid prop drilling when data is confined to one branch of the component hierarchy and does not require frequent updates or change detection.
+
+## Consider instrumenting and measuring render performance before and after recoil
+- MobX, Redux, and Recoil have different approaches to change detection and triggering rendering. 
+- Consider measuring the number and duration of render calls, compute function updates, and state updates.
+
+# How to implement a dispatcher pattern
+
+
+## Create a dispatcher composed of useRecoilCallback functions
+
+```typescript
+export const createDispatcher = () => {
+    // Example: Setting new state
+    // Supports frequent multiple calls sometimes from other callbacks.
+    const logMessage = useRecoilCallback<[string], void>((callbackHelpers: CallbackInterface, message: string) => {
+        const { set } = callbackHelpers;
+
+        // Example of using the set updator function to ensure latest state.
+        
+        set(logEntryListState, (logEntries) => [...logEntries, message]);
+    });
+
+    // Example: Reset state, load from file, and set state.
+    const loadDocument = useRecoilCallback<[string], void>(async (callbackHelpers: CallbackInterface, file: string) => {
+        const { set, reset } = callbackHelpers;
+
+        reset(stepState);
+        reset(labelListState);
+        reset(labelingAreaListState);
+        reset(predictionListState);
+
+        const newText = fsExtra.readFileSync(file);
+        set(labeledDocumentState, { location: file, name: path.basename(file), text: newText.toString() });
+
+        logMessage(`loaded document: ${file}`);
+    });
+
+    // Example: Using getPromise to read data, set new state, and then return a value
+    // Note that callers that are not part of a UI callback (i.e. onClick) need to await the Promise<Label> return value
+    const addLabel = useRecoilCallback<[Label, CharacterRange, Entity], Promise<Label>>(
+        async (callbackHelpers: CallbackInterface, parent: Label, charRange: CharacterRange, entity: Entity) => {
+            const { getPromise, set } = callbackHelpers;
+
+            const labels = await getPromise(labelListState);
+            const newLabel: Label = {
+                id: uuidv4(),
+                charRange: charRange,
+                entityId: entity.id,
+                parent: { kind: 'label', id: parent.id },
+            };
+            set(labelListState, [...labels, newLabel]);
+            logMessage(`+label ${entity.name}  ${characterRangeToString(newLabel.charRange)}`);
+            return newLabel;
+        }
+    );
+...
```

Comment:
Is this were we would put our actions? Like fetching data from the server, etc?

Reply:
Looks like the dispatchers are like factory functions. We make different dispatcher sets of functions for different purposes and leverage them from the components to update state. So yeah actions and reducers combined I guess.

Reply:
So the official documentation uses selectors to make those calls.
https://recoiljs.org/docs/guides/asynchronous-data-queries/